agent/agent.go

```diff
@@ -639,13 +641,77 @@ func (a *agent) runScript(ctx context.Context, lifecycle, script string) error {
 	}
 
 	a.logger.Info(ctx, "running script", slog.F("lifecycle", lifecycle), slog.F("script", script))
-	writer, err := a.filesystem.OpenFile(filepath.Join(a.logDir, fmt.Sprintf("coder-%s-script.log", lifecycle)), os.O_CREATE|os.O_RDWR, 0o600)
+	fileWriter, err := a.filesystem.OpenFile(filepath.Join(a.logDir, fmt.Sprintf("coder-%s-script.log", lifecycle)), os.O_CREATE|os.O_RDWR, 0o600)
 	if err != nil {
 		return xerrors.Errorf("open %s script log file: %w", lifecycle, err)
 	}
 	defer func() {
-		_ = writer.Close()
+		_ = fileWriter.Close()
 	}()
+
+	startupLogsReader, startupLogsWriter := io.Pipe()
+	defer func() {
+		_ = startupLogsReader.Close()
+		_ = startupLogsWriter.Close()
+	}()
+	writer := io.MultiWriter(startupLogsWriter, fileWriter)
+
+	queuedLogs := make([]agentsdk.StartupLog, 0)
+	var flushLogsTimer *time.Timer
+	var logMutex sync.Mutex
+	var logsSending bool
+	sendLogs := func() {
+		logMutex.Lock()
+		if logsSending {
+			logMutex.Unlock()
+			return
+		}
+		logsSending = true
+		logsToSend := queuedLogs
+		queuedLogs = make([]agentsdk.StartupLog, 0)
+		logMutex.Unlock()
+		for r := retry.New(time.Second, 5*time.Second); r.Wait(ctx); {
+			err := a.client.PatchStartupLogs(ctx, agentsdk.PatchStartupLogs{
+				Logs: logsToSend,
+			})
+			if err == nil {
+				break
+			}
+			a.logger.Error(ctx, "upload startup logs", slog.Error(err), slog.F("to_send", logsToSend))
+		}
+		logMutex.Lock()
+		logsSending = false
+		logMutex.Unlock()
+	}
+	queueLog := func(log agentsdk.StartupLog) {
+		logMutex.Lock()
+		defer logMutex.Unlock()
+		queuedLogs = append(queuedLogs, log)
+		if flushLogsTimer != nil {
+			flushLogsTimer.Reset(100 * time.Millisecond)
+			return
+		}
+		if len(queuedLogs) > 100 {
+			go sendLogs()
+			return
+		}
+		flushLogsTimer = time.AfterFunc(100*time.Millisecond, func() {
+			sendLogs()
+		})
+	}
+	err = a.trackConnGoroutine(func() {
+		scanner := bufio.NewScanner(startupLogsReader)
```

Comment:
given that we'll silently drop output after the first KB on the DB-side what if we modified the buffer size here to avoid dropping output? 
```golang
buffer := make([]byte, 0, 1<<10)
scanner.Buffer(buffer, 1<<10)
```

Reply:
Smart